Proof serialization/de-serialization (#155)
decentralized-identity/bbs-signature: draft-bbs-signatures.md

X\[a..b\]
: Denotes a slice of the array `X` containing all elements from and including the value at index `a` until and including the value at index `b`. Note when this syntax is applied to an octet string, each element in the array `X` is assumed to be a single byte.

> (a, ..., b)

Reviewer: ```suggestion
(a..b)
```

Reviewer: OR something like python `range(0,2)`

Author: Thanks! Updated to `range(0, 2)` (we use [a..b] for array/octet_string slices)

- octet_point_length (REQUIRED), non-negative integer. The length of
                                 a point in G1 octet representation,
                                 defined by the ciphersuite.

Outputs:

- proof, a proof value in the form described above or INVALID

Procedure:

1. proof_len_floor = 3 * octet_point_length + 5 * octet_scalar_length

2. if length(proof_octets) < proof_len_floor, return INVALID

// Points (i.e., (A', Abar, D) in ProofGen) de-serialization.
3. index = 0

4. for i in range(0, 2):

5.     end_index = index + octet_point_length - 1

6.     A_i = octets_to_point_g1(proof_octets[index..end_index])

7.     if A_i is INVALID or Identity_G1, return INVALID

8.     index += octet_point_length

// Scalars (i.e., (c, e^, r2^, r3^, s^, (m^_j1, ..., m^_jU)) in
// ProofGen) de-serialization.
9. j = 0

10. while index < length(proof_octets):

11.     end_index = index + octet_scalar_length - 1

12.     s_j = OS2IP(proof_octets[index..end_index])

13.     if s_j = 0 or if s_j >= r, return INVALID

14.     index += octet_scalar_length

15.     j += 1

16. if index is not equal to length(proof_octets), return INVALID

> 17. Let msg_commitments be an empty set (i.e., msg_commitments = ())

Reviewer: Otherwise from a data representation perspective serialized as JSON you could take the current syntax to mean

```
proof = {
   A_0: ....,
   .....
   msg_commitments: {
      s_5,
      ...
   }
}
```

Instead

```
proof = {
   A_0: ....,
   .....
   s_5,
   ...
}
```

> 17. Let msg_commitments be an empty set (i.e., msg_commitments = ())

Author: I think that in the rest of the spec we use it as the first JSON example you listed. For example, in [proofGen](https://identity.foundation/bbs-signature/draft-bbs-signatures.html#name-proofgen), in step 25 we compute,
```
25. proof = (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1, ..., m^_jU))
```
And in [ProofVerify](https://identity.foundation/bbs-signature/draft-bbs-signatures.html#name-proofverify), at step 4, we (currently) parse the proof as,
```
4. (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1,...,m^_jU)) = proof
```
Show the options I see is
1. We update this on those places as well (will be really easy to do in this PR)
2. We keep using the same notation

I'm fine with both. I don't think it make much difference from an implementations perspective. From an aesthetics perspective i may go with option 1, so there is some distinction between the different parts of the proof, but really not filling strong about it.

> 17. Let msg_commitments be an empty set (i.e., msg_commitments = ())

Reviewer: Im actually fine with either so long as we are consistent, suffice to say if you think we are already being consistent w.r.t  option 1 then we can leave it as is.

> 17. Let msg_commitments be an empty set (i.e., msg_commitments = ())

Reviewer: There is a pending suggested change attached to this comment also which isn't entirely related to this conversation, once we have a resolution on that I'm happy to approve

> 17. Let msg_commitments be an empty set (i.e., msg_commitments = ())

Author: Yeap! Updated based on suggestion. Thanks!

4. return (a_octets || e_octets || s_octets)
```

### OctetsToProof

> The `octets_to_proof` operations describes how to parse an octet string representation of a proof, validate it and return the underlying components.

Reviewer: ```suggestion
This operation describes how to decode an octet string representing a proof, validate it and return the underlying components that make up the proof.
```

Suggestion to make this more consistent language with how we introduce each operation
```suggestion
The `octets_to_proof` operations describes how to parse an octet string representation of a proof, validate it and return the underlying components.
```

Author: Done!

> The `octets_to_proof` operations describes how to parse an octet string representation of a proof, validate it and return the underlying components.

Reviewer: ```suggestion
This operation describes how to decode an octet string representing a proof, validate it and return the underlying components that make up the proof.
```

Suggestion to make this more consistent language with how we introduce each operation
```suggestion
The `octets_to_proof` operations describes how to parse an octet string representation of a proof, validate it and return the underlying components.
```

Author: Thank you! Updated

The proof value outputted by this operation consists of the following components, in that order:

> 1. Three (3) valid points of the G1 subgroup, different from the identity point.

Reviewer: ```suggestion
1. Three (3) valid points of the G1 subgroup, each of which must not equal the identity point.
```

> 1. Three (3) valid points of the G1 subgroup, different from the identity point.

Reviewer: ```suggestion
1. Three (3) valid points of the G1 subgroup, each of which must not equal the identity point.
```

Author: Thanks! Updated!

2. Five (5) integers representing scalars in the range of 1 to r-1 inclusive.

> 3. A number of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the un-revealed from the proof message commitments.

Reviewer: ```suggestion
3. A set of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the un-revealed from the proof message commitments.
```

Also do we need to clarify the ordering expected here?
```suggestion
3. A number of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the un-revealed from the proof message commitments.
```

Author: > Also do we need to clarify the ordering expected here?

There is no way that i know of for the ordering to be checked here, so there may be no point in adding such description?? The expected "ordering" is kinda "defined" in step 4 in `ProofVerify` i.e., 

```
4. (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1,...,m^_jU)) = proof
```
Since the functions are just utilities to be used internally and there are no inconsistencies that can arise when used with those core operations, an extra definition of the ordering may not be necessary??

> 3. A number of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the un-revealed from the proof message commitments.

Reviewer: ```suggestion
3. A set of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the un-revealed from the proof message commitments.
```

Also do we need to clarify the ordering expected here?
```suggestion
3. A number of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the un-revealed from the proof message commitments.
```

```
proof = octets_to_proof(proof_octets)

Inputs:

- proof_octets (REQUIRED), octet string of the form output from
                           proof_to_octets operation.

Parameters:

> - r (REQUIRED), non-negative integer. The prime order of the G_1 and
>                 G_2 groups, defined by the ciphersuite.

Reviewer: ```suggestion
- r (REQUIRED), non-negative integer. The prime order of the G1 and
                G2 groups, defined by the ciphersuite.
```
```suggestion
- r (REQUIRED), non-negative integer. The prime order of the G_1 and
                G_2 groups, defined by the ciphersuite.
```

> - r (REQUIRED), non-negative integer. The prime order of the G_1 and
>                 G_2 groups, defined by the ciphersuite.

Reviewer: ```suggestion
- r (REQUIRED), non-negative integer. The prime order of the G1 and
                G2 groups, defined by the ciphersuite.
```
```suggestion
- r (REQUIRED), non-negative integer. The prime order of the G_1 and
                G_2 groups, defined by the ciphersuite.
```

Author: Thanks! Updated!

> - octet_scalar_length (REQUIRED), non-negative integer. The length of

Reviewer: Just a note we should probably go back through the spec and make this consistent if we want to adopt a pattern of defining parameters that are set by the cipher suite? Shall I create an issue?

Author: Agreed! There is issue #125. We do need to focus more into closing it so the spec will be consistent.

> - octet_scalar_length (REQUIRED), non-negative integer. The length of

Reviewer: Just a note we should probably go back through the spec and make this consistent if we want to adopt a pattern of defining parameters that are set by the cipher suite? Shall I create an issue?

                                  a scalar octet representation, defined
                                  by the ciphersuite.
- octet_point_length (REQUIRED), non-negative integer. The length of
                                 a point in G1 octet representation,
                                 defined by the ciphersuite.

Outputs:

- proof, a proof value in the form described above or INVALID

Procedure:

1. proof_len_floor = 3 * octet_point_length + 5 * octet_scalar_length

2. if length(proof_octets) < proof_len_floor, return INVALID

// Points (i.e., (A', Abar, D) in ProofGen) de-serialization.
3. index = 0

> 4. for i in (0, 2):

Reviewer: Just checking this loop syntax is 0,1,2 wonder if we should do `(0,..,2)` and define the syntax in the terminology section?

Author: Yeap! Good catch! Updated this line and added the `(a, ..., b)` syntax in the Notation section.

> 4. for i in (0, 2):

Reviewer: Just checking this loop syntax is 0,1,2 wonder if we should do `(0,..,2)` and define the syntax in the terminology section?

Author: Yeap! Done!

5.     end_index = index + octet_point_length - 1

6.     A_i = octets_to_point_g1(proof_octets[index..end_index])

7.     if A_i is INVALID or Identity_G1, return INVALID

8.     index += octet_point_length

// Scalars (i.e., (c, e^, r2^, r3^, s^, (m^_j1, ..., m^_jU)) in
// ProofGen) de-serialization.
9. j = 0

10. while index < length(proof_octets):

11.     end_index = index + octet_scalar_length - 1

12.     s_j = OS2IP(proof_octets[index..end_index])

13.     if s_j = 0 or if s_j >= r, return INVALID

14.     index += octet_scalar_length

15.     j += 1

16. if index is not equal to length(proof_octets), return INVALID

> 17. msg_commitments = (s_5, ..., s_(k-1))

Reviewer: Where is k defined?

Author: Apologies! This was a typo. It should be `j`. Updated.

> 17. msg_commitments = (s_5, ..., s_(k-1))

Reviewer: Where is k defined?

18. return (A_0, A_1, A_2, s_0, s_1, s_2, s_3, s_4, msg_commitments)
```

### ProofToOctets

This operation describes how to encode a proof, as computed at step 25 in [ProofGen](#proofgen), to an octet string. The input to the operation MUST be a valid proof.

The inputed proof value must consist of the following components, in that order:

> 1. Three (3) valid points of the G1 subgroup, different from the identity point of G1 (i.e., `A', Abar, D`, in ProofGen)

Reviewer 2: ```suggestion
1. Three (3) valid compressed points of the G1 subgroup, different from the identity point of G1 (i.e., `A', Abar, D`, in ProofGen)
```

Author: Will the points be compressed when calculated in ProofGen?? Wouldn't they be just regular points (not octets outputted by `point_to_octets`)??

2. Five (5) integers representing scalars in the range of 1 to r-1 inclusive (i.e., `c, e^, r2^, r3^, s^`, in ProofGen).

> 3. A number of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the un-revealed from the proof messages (i.e., `m^_j1, ..., m^_jU`, in ProofGen, where U the number of un-revealed messages).

Reviewer 2: Little endian 32-bit? Can’t remember if we have a description of serialized integers yet

Author: Thank you! I don't think we discuss this on the document. Aren't there 256 bits necessary for scalars (for bls12-381)?? Also we use I2OSP/OS2IP which operates in big-endian order if I'm not mistaken, so we will need to add a note there if we decide to define integers to be little endian.

```
proof_octets = proof_to_octets(proof)

Inputs:

> - proof (REQUIRED), a BBS proof in the form calculated by ProofGen in

Reviewer: Wonder if we actually need to supply the arguments destructed here like we do in signature to octets?

> - proof (REQUIRED), a BBS proof in the form calculated by ProofGen in

Reviewer: Wonder if we actually need to supply the arguments destructed here like we do in signature to octets?

Author: Since the input has no "fixed" length it may be more complicated to define it ''destructed" (i.e., if we define it as `(A', Abar, D, c, e^, r2^, r3^, s^, (m^_1,...,m^_U))` we will need to define `U` which will be hard here or if we define it like `(A', Abar, D, c, e^, r2^, r3^, s^, msg_commitments`) we mix up structured and destructed forms which misses the point of consistency IMO.

In general, for consistency I would prefer passing the signature in a structured way than passing the proof in a destructed one.

Reviewer: Happy to move in that direction do we need to raise an issue to capture that?

Author: Updated this PR to make the change for `octets_to_signature` and `signature_to_octets` (+ took the opportunity to make the operations < 72 chars). For other operations the update will most likely happen in #138. Opened #163 to keep track of this.

                    step 25 (see above).

Parameters:

- octet_scalar_length (REQUIRED), non-negative integer. The length of
                                  a scalar octet representation, defined
                                  by the ciphersuite.

Outputs:

- proof_octets, octet string.

Procedure:

1. (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1, ..., m^_jU)) = proof

> 2. Let proof_octets be the empty octet string.

Reviewer: ```suggestion
2. Let proof_octets be an empty octet string.
```
```suggestion
2. Let proof_octets be the empty octet string.
```

> 2. Let proof_octets be the empty octet string.

Reviewer: ```suggestion
2. Let proof_octets be an empty octet string.
```
```suggestion
2. Let proof_octets be the empty octet string.
```
